src/workerd/util/state-machine.h

```diff
@@ -0,0 +1,3897 @@
+  StateMachine(const StateMachine&) = delete;
+  StateMachine& operator=(const StateMachine&) = delete;
+
+  // Initialize with a specific state
+  template <typename S, typename... Args>
+  explicit StateMachine(kj::Badge<S>, Args&&... args)
+    requires(_::isOneOf<S, States...>)
+  {
+    state.template init<S>(kj::fwd<Args>(args)...);
+  }
+
+  // Factory function for clearer initialization
+  template <typename S, typename... Args>
+  static StateMachine create(Args&&... args)
+    requires(_::isOneOf<S, States...>)
+  {
+    StateMachine m;
+    m.state.template init<S>(kj::fwd<Args>(args)...);
+    return m;
+  }
+
+  // ---------------------------------------------------------------------------
+  // Transition Locking
+  // ---------------------------------------------------------------------------
+
+  // Check if transitions are currently locked
+  bool isTransitionLocked() const {
+    return transitionLockCount > 0;
+  }
+
+  // Lock transitions (called by TransitionLock RAII guard)
+  void lockTransitions() {
+    ++transitionLockCount;
+  }
+
+  // Unlock transitions (called by TransitionLock RAII guard)
+  void unlockTransitions() {
+    KJ_DASSERT(transitionLockCount > 0, "Transition lock underflow");
+    --transitionLockCount;
+  }
+
+  // Get a RAII lock that prevents transitions while in scope
+  TransitionLock<StateMachine> acquireTransitionLock() {
+    return TransitionLock<StateMachine>(*this);
+  }
+
+  // ---------------------------------------------------------------------------
+  // State Queries
+  // ---------------------------------------------------------------------------
+
+  // Check if the machine is in a specific state.
+  // Returns false if uninitialized.
+  template <typename S>
+  bool is() const
+    requires(_::isOneOf<S, States...>)
+  {
+    return state.template is<S>();
+  }
+
+  // Check if the machine is in any of the specified states.
+  // Returns false if uninitialized.
+  template <typename... Ss>
+  bool isAnyOf() const
+    requires((_::isOneOf<Ss, States...>) && ...)
+  {
+    return (is<Ss>() || ...);
+  }
+
+  // Check if the machine is initialized (not in the null state).
+  // Call transitionTo<>() to initialize the state machine.
+  bool isInitialized() const {
+    return !(state == nullptr);
+  }
+
+  // Assert that the machine is initialized, with a clear error message.
+  void requireInitialized() const {
+    KJ_REQUIRE(isInitialized(),
+        "State machine used before initialization. Call transitionTo<InitialState>() first.");
+  }
+
+  // ---------------------------------------------------------------------------
+  // State Access
+  // ---------------------------------------------------------------------------
+  //
+  // WARNING: get() and tryGet() return UNLOCKED references. The returned
+  // reference becomes invalid if the state machine transitions. Prefer
+  // withState() for safe access that locks transitions during the callback.
```

Comment:
I wonder if we could/should name these `getUnsafe()` and `tryGetUnsafe()` because of this concern? It would force callers to think harder.